pkg/providers/snow/defaults.go

```diff
@@ -0,0 +1,67 @@
+package snow
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/aws"
+	"github.com/aws/eks-anywhere/pkg/filewriter"
+	"github.com/aws/eks-anywhere/pkg/logger"
+)
+
+type Defaulters struct {
+	awsClients aws.Clients
+	writer     filewriter.FileWriter
+}
+
+func NewDefaulters(aws aws.Clients, writer filewriter.FileWriter) *Defaulters {
+	return &Defaulters{
+		awsClients: aws,
+		writer:     writer,
+	}
+}
+
+func (d *Defaulters) setDefaultSshKey(ctx context.Context, m *v1alpha1.SnowMachineConfig) (keyName, keyVal string, err error) {
+	if m.Spec.SshKeyName != "" {
+		return "", "", nil
+	}
+
+	keyName = fmt.Sprintf("%s-%s.pem", defaultAwsSshKeyNamePrefix, m.GetName())
+
+	logger.V(1).Info("SnowMachineConfig SshKey is empty. Creating default key pair", "default key name", keyName)
```

Comment:
I'm still not sure this is adequate for associating a given key with a particular machine config.

If I have 3 different machine configurations - CP, worker and etcd - how do I know which key belongs to which machine config and therefore which machine?

Moreover, this doesn't compliment other provider behavior where it re-uses the same key across all machine configs missing ssh keys. I battled with how to perform this logic in Tinkerbell because it wasn't readily re-usable from the vSphere provider. In the end, I followed an approximate flow of: (1) find machine configs without keys; (2) if we have machine configs without keys generate a key and tell the user where it is in a machine config agnostic manner; (3) apply the key to machine configs. It was a substantial simplification on the vSphere approach which is extremely convoluted.

Amittedly this all needs reworking so we have reusable defaulting logic but I still think there's some changes required here to meet the same behavior and be easy for the end user.

Reply:
this makes sense. i agree we should do the check and only generate one ssh key for all machines. I will have a separate PR for this and work on it next if you are ok. It is a feature improvement